orca-core/src/main/java/com/netflix/spinnaker/orca/locks/LockContext.java

```diff
@@ -0,0 +1,139 @@
+      private String type;
+      private String id;
+      private Stage stage;
+
+      LockValueBuilder() {
+        this(null, null, null, null);
+      }
+
+      @JsonCreator
+      public LockValueBuilder(@JsonProperty("application") String application,
+                              @JsonProperty("type") String type,
+                              @JsonProperty("id") String id) {
+        this(application, type, id, null);
+      }
+
+      LockValueBuilder(String application, String type, String id, Stage stage) {
+        this.application = application;
+        this.type = type;
+        this.id = id;
+        this.stage = stage;
+      }
+
+      public LockValueBuilder withStage(Stage stage) {
+        this.stage = stage;
+        return this;
+      }
+
+      public LockManager.LockValue build() {
+        final Optional<Execution> execution = Optional.ofNullable(stage).map(Stage::getExecution);
+
+        final String application = Optional.ofNullable(this.application).orElseGet(() ->
+          execution
+            .map(Execution::getApplication)
+            .orElse(null));
+
+        final String type = Optional.ofNullable(this.type).orElseGet(() ->
+          execution
+            .map(Execution::getType)
+            .map(Execution.ExecutionType::toString)
+            .orElse(null));
+
+        final String id = Optional.ofNullable(this.id).orElseGet(() ->
+          execution
+            .map(Execution::getId)
+            .orElse(null));
+
+        return new LockManager.LockValue(application, type, id);
+      }
+    }
+
+
+    private String lockName;
+    private LockValueBuilder lockValue;
+    private String lockHolder;
+    private Stage stage;
+
+    public LockContextBuilder() {
+      this(null, null, null, null);
+    }
+
+    @JsonCreator
+    public LockContextBuilder(
+      @JsonProperty("lockName") String lockName,
+      @JsonProperty("lockValue") LockValueBuilder lockValue,
+      @JsonProperty("lockHolder") String lockHolder) {
+      this(lockName, lockValue, lockHolder, null);
+    }
+
+    LockContextBuilder(String lockName, LockValueBuilder lockValue, String lockHolder, Stage stage) {
+      this.lockName = lockName;
+      this.lockValue = lockValue;
+      this.lockHolder = lockHolder;
+      this.stage = stage;
+    }
+
+    public LockContextBuilder withStage(Stage stage) {
+      this.stage = stage;
+      Optional.ofNullable(lockValue).ifPresent(lv -> lv.withStage(stage));
+      return this;
+    }
+
+    public LockContext build() {
+      final LockManager.LockValue lockValue =
+        Optional.ofNullable(this.lockValue)
+          .orElseGet(() -> new LockValueBuilder().withStage(stage))
+          .build();
+
+      //TODO(cfieber):
+      //lockHolder - if we are a synthetic stage then should we use parent stage's id?
+      // if we are a top level stage, lock with our id
+      // the corresponding release lock stage should have an explicit stage id in it..
```

Comment:
If we're a synthetic stage in say... a canary or something new and exciting, we'd want locks on individual synthetic deploy stages yes?

Reply:
... was typing the same thing as @robzienert. 

The individual synthetic deploy stages have the execution windows attached (pretty sure?), so we'd want to lock at that level.

But there are probably more normal cases where we'd want to use the parent stage id.

Reply:
in hindsight I think this comment is kind of unnecessary - the holder is just an id for one (of potentially many) holders of a lock - used to enable the reentrant locking bits. I think I need to move this comment into the LockValueBuilder where it is figuring out the lockValue (which will be the execution.id by default). There we need a story for locking in child pipelines or strategies.